tsl/src/compression/compression_dml.c

```diff
@@ -346,44 +315,117 @@ decompress_batches_for_update_delete(HypertableModifyState *ht_state, Chunk *chu
+						   ScanKeyData *scankeys)
+{
+	DecompressBatchScanDesc scan;
+	scan = (DecompressBatchScanDesc) palloc(sizeof(DecompressBatchScanData));
+
+	if (index_rel)
+	{
+		scan->index_scan = index_beginscan(in_rel, index_rel, snapshot, num_scankeys, 0);
+		index_rescan(scan->index_scan, scankeys, num_scankeys, NULL, 0);
+		scan->scan = NULL;
+	}
+	else
+	{
+		scan->scan = table_beginscan(in_rel, snapshot, num_scankeys, scankeys);
+		scan->index_scan = NULL;
+	}
+
+	return scan;
+}
+
+static bool
+decompress_batch_scan_getnext_slot(DecompressBatchScanDesc scan, ScanDirection direction,
+								   struct TupleTableSlot *slot)
+{
+	if (scan->index_scan)
+	{
+		return index_getnext_slot(scan->index_scan, direction, slot);
+	}
+	else
+	{
+		return table_scan_getnextslot(scan->scan, direction, slot);
+	}
+}
+
+static void
+decompress_batch_endscan(DecompressBatchScanDesc scan)
+{
+	if (scan->index_scan)
+	{
+		index_endscan(scan->index_scan);
+	}
+	else
+	{
+		table_endscan(scan->scan);
+	}
+
+	pfree(scan);
+}
+
 /*
  * This method will:
- *  1.Scan the index created with SEGMENT BY columns.
+ *  1.Scan the index created with SEGMENT BY columns or the entire compressed chunk
  *  2.Fetch matching rows and decompress the row
- *  3.insert decompressed rows to uncompressed chunk
- *  4.delete this row from compressed chunk
+ *  3.Delete this row from compressed chunk
+ *  4.Insert decompressed rows to uncompressed chunk
  *
  *  Returns whether we decompressed anything.
+ *
  */
 static struct decompress_batches_stats
-decompress_batches_indexscan(Relation in_rel, Relation out_rel, Relation index_rel,
-							 Snapshot snapshot, ScanKeyData *index_scankeys, int num_index_scankeys,
-							 ScanKeyData *heap_scankeys, int num_heap_scankeys,
-							 ScanKeyData *mem_scankeys, int num_mem_scankeys,
-							 tuple_filtering_constraints *constraints, bool *skip_current_tuple,
-							 bool delete_only, Bitmapset *null_columns, List *is_nulls)
+decompress_batches_scan(Relation in_rel, Relation out_rel, Relation index_rel, Snapshot snapshot,
+						ScanKeyData *index_scankeys, int num_index_scankeys,
+						ScanKeyData *heap_scankeys, int num_heap_scankeys,
+						ScanKeyData *mem_scankeys, int num_mem_scankeys,
+						tuple_filtering_constraints *constraints, bool *skip_current_tuple,
+						bool delete_only, Bitmapset *null_columns, List *is_nulls)
 {
 	HeapTuple compressed_tuple;
 	RowDecompressor decompressor;
 	bool decompressor_initialized = false;
 	bool valid = false;
-	int num_segmentby_filtered_rows = 0;
-	int num_heap_filtered_rows = 0;
+	int num_scanned_rows = 0;
+	int num_filtered_rows = 0;
+	TM_Result result;
+	DecompressBatchScanDesc scan = NULL;
 
 	struct decompress_batches_stats stats = { 0 };
 
-	/* TODO: Optimization by reusing the index scan while working on a single chunk */
-	IndexScanDesc scan = index_beginscan(in_rel, index_rel, snapshot, num_index_scankeys, 0);
+	if (index_rel)
```

Comment:
One of the reasons why I would like to have the scan preparation outside of this function is because you have this branching in the decompress batch scan already, making this look superfluous here.

Reply:
Alright, let's take this up in a follow-up PR.